Search with regex

jsquaredosquared opened the pull request:
Hi @nnako, thanks for this incredibly useful package.

I have a suggestion that the search could be performed using regular expressions. I think it would allow some more powerful searches and simplify some of the code, for example:

```python
mindmap = freeplane.Mindmap("example.mm")

# The `(?i)` regex flag makes the search case insensitive.
# This will match "test" and "TEsTer" but not "teSTinG".
mindmap.find_nodes(core=r"(?i)test\w{,2}\b")

# This finds all nodes where the value of the "Scheduled" attribute begins with 2025.
mindmap.find_nodes(attrib={"Scheduled": r"^2025"})
```

What do you think?

jsquaredosquared committed:
feat!: perform searching using regular expressions

jsquaredosquared committed:
style: revert inverted comma changes

nnako commented:
Hi @jsquaredosquared ,

thank you very much for your useful suggestion. Regular Expressions provide powerful search capabilities. As I don't want to break compatibility, I'd rather keep the default search concept as is. But we could add a `regex=<bool>` parameter into all the `find` functions like:

```python
<mindmap>.find_nodes()
<node>.find_nodes()
<node>.find_children()
```

so that examples could look like this:

```python
mindmap = freeplane.Mindmap("example.mm")

# The `(?i)` regex flag makes the search case insensitive.
# This will match "test" and "TEsTer" but not "teSTinG".
mindmap.find_nodes(core=r"(?i)test\w{,2}\b", regex=True)

# This finds all nodes where the value of the "Scheduled" attribute begins with 2025.
mindmap.find_nodes(attrib={"Scheduled": r"^2025"}, regex=True)
```

When using `regex=True`, the other parameters like

```python
exact=<bool>
caseinsensitive=<bool>
generalpathsep=<bool>
keep_link_specials=<bool>
```

would have no effect. The user of `regex=<bool>` should be able to use Regular Expressions, accordingly.

What do you think?

jsquaredosquared committed:
fix: keep existing behaviour, add regex kwarg

jsquaredosquared committed:
style: revert formatter changes

jsquaredosquared commented:
Hey @nnako, I completely understand not wanting to break compatibility.

I have rewritten the code to go back to the previous search concept and added a regex parameter as you suggested.

For example:
```python
if (
    (regex and re.search(details, _text))
    or (exact and not caseinsensitive and details == _text)
    or (exact and caseinsensitive and details.lower() == _text.lower())
    or (not exact and details.lower() in _text.lower())
):
    _lstNodes.append(_node)

```

This should keep the previous behaviour, although perhaps some testing is needed to make sure :D

jsquaredosquared committed:
fix: searching in details

jsquaredosquared committed:
fix: searching in details

jsquaredosquared committed:
fix: prevent returning nodes matching attributes more than once

nnako commented:
Hi @jsquaredosquared , 

I'd like to mention you as a Co-Author within the commit history of this project, but I was not able to determine your respective / preferred email address. Would you be so kind and tell me (e.g. via one of my email addresses) which email you want to appear within the commit history?

nnako commented:
Thanks again, @jsquaredosquared , for your contribution.
